### ISSUE TITLE
Pocketbook 633 Color support

### ISSUE THREAD

**Issue by axispod** [koreader/koreader#6479]
* KOReader version: 2020.07.1
* Device: Pocketbook 633 Color

#### Issue
I have installed it on my new PB 633 Color.
But the latest version of KOReader crashes on startup

#### Steps to reproduce
Just run

##### `crash.log` (if applicable)

```
---------------------------------------------
                launching...
  _  _____  ____                _
 | |/ / _ \|  _ \ ___  __ _  __| | ___ _ __
 | ' / | | | |_) / _ \/ _` |/ _` |/ _ \ '__|
 | . \ |_| |  _ <  __/ (_| | (_| |  __/ |
 |_|\_\___/|_| \_\___|\__,_|\__,_|\___|_|

 It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 08/06/20-18:33:44
 [*] Version: v2020.07.1

ffi.load: SDL2
ffi.load (warning): libSDL2.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so
ffi.load (warning): libSDL2-2.0.so: cannot open shared object file: No such file or directory
ffi.load: libSDL2-2.0.so.0
ffi.load (warning): libSDL2-2.0.so.0: cannot open shared object file: No such file or directory
SDL2 not loaded:	./setupkoenv.lua:27: Not able to load dynamic library: libSDL2-2.0.so.0
ffi.load: inkview
08/06/20-18:33:44 INFO  SoftwareVersion:  U633.6.1.488
08/06/20-18:33:44 INFO  Blacklisting the C BB on this device
Tweaking LuaJIT's max loop unroll factor (15 -> 45)
PB FB: smem_len    : 1552384
PB FB: line_length : 1448
PB FB: xres        : 1448
PB FB: xres_virtual: 1448
PB FB: yres        : 1072
PB FB: yres_virtual: 1072
PB FB: bpp         : 8
PB FB: xres_virtual -> 1472
PB FB: yres_virtual -> 1152
PB FB: line_length -> 1472
PB FB: yres_virtual => 1088
PB FB: line_length <- 1448
PB FB: xres_virtual <- 1448
PB FB: yres_virtual <- 1072
FB: computed size: 1552384
FB: id           : 
FB: smem_len     : 1552384
FB: line_length  : 1448
FB: xres         : 482
FB: xres_virtual : 1448
FB: yres         : 1072
FB: yres_virtual : 1072
FB: bpp          : 24
Segmentation fault 
```

**Comment by Frenzie** [comment]
Segmentation fault? That's disquieting.

**Comment by NiLuJe** [comment]
Probably a duplicate of #6000 and/or related to it (i.e., FW 6.x InkView issues).

**Comment by Frenzie** [comment]
Good point, I forgot about that.

**Comment by roshavagarga** [comment]
@axispod Can you check what the 'model' file says in your device's system directory?

**Comment by Teteros** [comment]
No 'model' file on my PocketBook Color at least, anything else that might help?

**Comment by roshavagarga** [comment]
@Teteros It's in the system directory actually, which is hidden by default.

**Comment by Teteros** [comment]
Yes, I'm looking at system dir from USB mount point, here's a dump of what the filetree contains:
[systemdir.txt](https://github.com/koreader/koreader/files/5045608/systemdir.txt)

**Comment by roshavagarga** [comment]
@Teteros Install koreader 2020.07.1, but replace `koreader/frontend/device/pocketbook/device.lua` with this one: https://github.com/roshavagarga/koreader/blob/master/frontend/device/pocketbook/device.lua

Even if this one doesn't help the device work, it'll give the developers a better crash log than the previous one, or the appropriate codename, since if the one I placed is incorrect, Koreader will spew out the appropriate one :+1:

**Comment by Teteros** [comment]
Seems the crash.log with this new device.lua isn't any different than the original one in this issue :/

**Comment by roshavagarga** [comment]
@Teteros Is it exactly the same?

**Comment by Teteros** [comment]
Yep, ran a diff and only the timestamps are different.

**Comment by roshavagarga** [comment]
Welp, I'll leave it up to the real devs, seems like the definition wasn't helpful.

**Comment by Teteros** [comment]
I don't know how helpful this is because the device isn't actually rooted, but I managed to get strace on koreader.app using [this](https://www.mobileread.com/forums/showthread.php?t=116350) script which is more info than the crash.log at least.
[koreader_2020.07.1_pb633_strace.log](https://github.com/koreader/koreader/files/5045838/koreader_2020.07.1_pb633_strace.log)

**Comment by roshavagarga** [comment]
That might be useful, yes, @Frenzie might be able to make heads or tails of it, but the permission denied at the end is suspicious.

**Comment by Frenzie** [comment]
@roshavagarga A segmentation fault isn't caused by anything in the device definition. It's likely similar to #6000 as @NiLuJe said, except it'll be an even more basic library. See there for how to get a more useful gdb backtrace.

Perhaps some kind of libinkview wrapper could work around the problem?

**Comment by NiLuJe** [comment]
It'd probably need to be in a different process space, which makes this tricky, as we need it for input, among other things...

**Comment by Teteros** [comment]
@Frenzie Thanks, managed to built koreader-master in debug mode using the [koreader/kopb:0.2.0](https://hub.docker.com/r/koreader/kopb/tags) docker image.
Here's the gdb backtrace koreader spits out on the pocketbook:
[gdb.txt](https://github.com/koreader/koreader/files/5051639/gdb.txt)

**Comment by NiLuJe** [comment]
Huh. That's... interesting :D.

It's face-planting on a `memset` (via `ffi.fill`) from the `device` module...
The only thing that does explicit fills this way is the drawing code (i.e., blitbuffer).

Which probably means it's an OOB write on the fb, which would point to the hoop-jumping done in `framebuffer_linux.lua` to handle legacy PB devices...

I'd have said "try the C BB" (that might have given us a better backtrace), but you can't, because it's a Color device with an horribly broken fb setup.

**Comment by NiLuJe** [comment]
In fact, yeah, duh:

`FB: xres         : 482`

That's obviously wrong.

Thing is, that doesn't look anything at all like the logs from the original implementation.

Is this a new revision? What's the actual effective resolution of this thing?

**Comment by NiLuJe** [comment]
Which possibly makes sense, given that https://github.com/koreader/koreader/pull/4212 was for a Lux, not a 633 ;).

So, yeah, a specs sheet would be nice.

**Comment by Frenzie** [comment]
To be clear, has anybody tried the obvious of *not* including `has3BytesWideFrameBuffer` to see what happens? I would've gone the opposite way from what @pazos suggested (not include it, try it if that doesn't work) but regardless what you start with… ;-)

**Comment by NiLuJe** [comment]
Yeah, that's a good first step.

If stuff renders in a corner or striped or weird, but "works", then we need another cap quirk, where we essentially do a * 3 instead of a / 3.... *sigh*.

**Comment by roshavagarga** [comment]
@NiLuJe Failed to find a manual or firmware update, so the only thing I can offer is the basic info available [here](https://pocketbook.ch/en-ch/catalog/color/color-ch).

**Comment by NiLuJe** [comment]
Yeah, that's the best I could find, too, unfortunately (as, it's, err, fairly useless. It just confirms that it's eInk's latest filter, probably running on an i.MX7D, which doesn't really help us ^^).

**Comment by roshavagarga** [comment]
@axispod @Teteros Would you mind attaching or writing out the specifications from your manuals?

**Comment by Teteros** [comment]
@roshavagarga @NiLuJe The manual and pictures for most of the internal components of this device can be found hosted on the FCC site which might help.
https://fcc.report/FCC-ID/2AUVWPB633

The manual included on the device is a bit prettier. Obviously a newer, better typeset version but specs are the same I believe.

**Comment by NiLuJe** [comment]
Yeah, that doesn't answer the effective pixelformat/resolution to handle color ;p.

But it does confirm that PB is sticking to their god-awful horribly outdated kernels, hence the misery.

**Comment by NiLuJe** [comment]
TL;DR: Someone with the actual device will just have to try what we mentioned earlier and see what happens ;).

**Comment by Frenzie** [comment]
Which Linux kernel does current Kobo run on?

**Comment by roshavagarga** [comment]
@NiLuJe One of those [pdfs](https://fcc.report/FCC-ID/2AUVWPB633/4783830.pdf) has photos of the internals. I was able to make out the following:

CPU: Allwinner B288 (KA280CA 62G1)
Wi-Fi?: Realtek (3761ATV G?14)

**Comment by Teteros** [comment]
Yep, just checked for you, on the latest firmware version U633.6.1.488 uname -a still prints 3.10.65
`Linux pocketbook 3.10.65+ #2 SMP Wed Jul 1 15:13:26 EEST 2020 armv7l GNU/Linux`

> TL;DR: Someone with the actual device will just have to try what we mentioned earlier and see what happens ;).

I've commented out the `has3BytesWideFrameBuffer` for PB633 in the definition file and KOReader actually runs now rather than crashes so big progress!
![CSC_0113](https://user-images.githubusercontent.com/5067989/90638556-61c21c80-e225-11ea-8754-fc18335ac073.JPG)
However there's something wrong with touch controls, the menus aren't responsive at all and sliding screen turns pages seemingly randomly, koreader might have the wrong idea of this devices resolution indeed.

**Comment by NiLuJe** [comment]
@Frenzie: 4.1.15. rM is ahead with the NXP drop following that one, on 4.9.something.

@roshavagarga: Gah. AW. That might explain the shitty kernel.  And also why everything is broken. We don't currently handle AW's EPDC.

**Comment by NiLuJe** [comment]
Are the kernel sources for this thing available somewhere?

**Comment by Frenzie** [comment]
@Teteros A typical scenario (at least on Kobo) is that touch is the wrong way around (i.e., top is bottom).

**Comment by NiLuJe** [comment]
As for the orientation thing, we have a `isAlwaysPortrait` device cap (or something, check the actual name/syntax :D) to handle those.

**Comment by Teteros** [comment]
> @Teteros A typical scenario (at least on Kobo) is that touch is the wrong way around (i.e., top is bottom).
You're right it seems to be indeed any way to reverse it?

**Comment by axispod** [comment]
> However there's something wrong with touch controls, the menus aren't responsive at all and sliding screen turns pages seemingly randomly

Just have added
```isAlwaysPortrait = yes```
And the control works properly now

**Comment by Frenzie** [comment]
@axispod Want to stick it in a PR? ^_^

**Comment by NiLuJe** [comment]
That still leaves Color to figure out, but, yeah, working grayscale instead of a crash would be a good start nonetheless ^^.

**Comment by roshavagarga** [comment]
@NiLuJe I think the PocketBook InkPad 3/3 Pro has the same CPU, but I might be wrong.

Links:
- [PocketBook open source github](https://github.com/pocketbook-free?tab=repositories) - last updated 6 years ago.
- [Ancient PocketBook modding](http://pocketbook-free.sourceforge.net/sourcecode.shtml) (kernel 2.X, on the 3XX devices) - last updated in 2015.
- [PocketBook github](https://github.com/pocketbook?tab=repositories) - last updated a year ago.

**Comment by Frenzie** [comment]
@NiLuJe Can't tell if there's no color based on the screenshot. ;-P

**Comment by roshavagarga** [comment]
@NiLuJe Best I could find is this, besides the older links up top - http://download.pocketbook-int.com/TestFW/SDK/toolchain-SDK-B288.tar.gz

**Comment by NiLuJe** [comment]
That simply appears to be a mildly updated variant of https://github.com/pocketbook/SDK_6.3.0/tree/5.19/SDK-B288...

That GH org ships the kernels in the Platform_<plat> repos, but, of course, it's missing the B288 one...

(Yum, GPL violation :D).

**Comment by roshavagarga** [comment]
@NiLuJe I think their ToS mentioned something along the lines of making source code available within 3 years when asked for, if not available, but their company is based in Switzerland, so if somebody actually asks and they don't comply they'd be in for a fun amount of issues :1st_place_medal:

**Comment by Teteros** [comment]
@Frenzie There isn't any no color rendering in any images or .pdf/.epub's I tried even though color is enabled in the options so that's not supported yet indeed.

**Comment by roshavagarga** [comment]
@NiLuJe [This review](https://goodereader.com/blog/reviews/pocketbook-color-e-reader-hands-on-review) mentiones that grayscale runs at 300 PPI, while the colored layer runs at 100 PPI. Photos of the device showing that difference are available [here](https://the-digital-reader.com/2020/08/05/hands-on-with-the-pocketbook-color-part-ii/).

**Comment by Teteros** [comment]
@roshavagarga color DPI might be closer to 173 than 100 if being pedantic (300/√3), see this thread which has more pictures https://www.mobileread.com/forums/showthread.php?t=331888&page=5

**Comment by axispod** [comment]
I found a news.
https://the-digital-reader.com/2020/04/15/pocketbook-to-launch-color-ereader-later-this-year-will-amazon-turn-green-with-envy/

It says the color PPI is 100.

When I discover a new screen using microscope, I found out that it have additional layer with blocks 2x2 where only 2 pixels have a color filters (3 vertical lines) in checkboard order. So DPI depends on hardware and I suppose it is 212ppi or 106ppi.

**Comment by axispod** [comment]
When I discover a new screen using microscope, I found out that it have additional layer with blocks 2x2 where only 2 pixels have a color filters (3 vertical lines) in checkboard order. I think the hardware treats the block a 1x2 or 2x1. So it has a half horizontal or vertical color resolution and it is 212ppi.

**Comment by axispod** [comment]
But also the hardware can work with block as one pixel in this case the resolution will be 150ppi.

**Comment by Frenzie** [comment]
@Teteros 
> color DPI might be closer to 173 than 100 if being pedantic (300/√3)

Sounds like a marketing fail if you're right?

**Comment by roshavagarga** [comment]
Either the marketing team had zero idea of the underlying technology or they decided to dumb it down for the users and not risk it being flagged as false advertisement.
Would the rather unique pixel layout be an issue for implementation?

**Comment by Teteros** [comment]
I've tried forcing 32bpp just in case but it segfaults on start as expected I suppose.
Guess this is a real 24bpp device unlike the Color Lux.
If anyone has any ideas throw them at me, perhaps FBGrab suggested [there](https://github.com/koreader/koreader/issues/4193#issuecomment-417965776) could be ported and we'd learn something, I wouldn't know :smile:

**Comment by NiLuJe** [comment]
As far as FBGrab is concerned, I doubt it'd give us much more info than currently, but, in any case, the [K3 binary](https://www.mobileread.com/forums/showpost.php?p=2854326&postcount=20) *should* run on PB.

`fbgrab -v /dev/null` should spit out all the nasty details.

But... this is PB: the fb setup usually *lies*.

In addition to that, no-one seems to know how the tech actually works, and I don't see how you could even request colors on a grayscale 8bpp fb o_O.

Either there's a magic flip somewhere that changes the fb layout for color, or there's another fb device/screen for color, or there's a custom fancy ioctl involved to request color refresh (think alt_buffer in i.MX land), or... something.
That still doesn't really answer how grayscale and color manage to get thrown together, which is apparently what happens on stock?

**Comment by Teteros** [comment]
@NiLuJe It did run and spit out this:
```
Fixed framebuffer info:
id: ""
start of fb mem: 0x7b600000
length of fb mem: 1552384 bytes
type: packed pixels
interleave for interleaved planes: 0
visual: true color
hw panning x step: 1
hw panning y step: 1
hw y wrap: 0
line length: 1448 bytes (1448 pixels)
start of mmio: 0
length of mmio: 0
accel chip/card: 0

Variable framebuffer info:
visible resolution: 1448x1072
virtual resolution: 1448x1072
offset from virtual to visible resolution: 0x0
bits per pixel: 8
grayscale: true
red:   bitfield offset: 5, bitfield length: 3, MSB is right: no
green: bitfield offset: 2, bitfield length: 3, MSB is right: no
blue:  bitfield offset: 0, bitfield length: 2, MSB is right: no
alpha: bitfield offset: 0, bitfield length: 0, MSB is right: no
pixel format: standard
activate: 128
height: 0 mm
width: 0 mm
obsolete accel_flags: 0
pixel clock: 0 ps
time from sync to picture (left margin): 0 pixclocks
time from picture to sync (right margin): 0 pixclocks
time from sync to picture (upper margin): 0 pixclocks
time from picture to sync (lower margin): 0 pixclocks
length of horizontal sync: 0 pixclocks
length of vertical sync: 0 pixclocks
sync: undefined! (0)
vmode: non interlaced
rotate: 2

Color map info:
start: 0, len: 256
  N      #R      #G      #B
00-03   0000    AAAA    AAAA
04-07   AAAA    AAAA    AAAA
08-0B   5555    FFFF    FFFF
0C-0F   FFFF    FFFF    FFFF
10-FF   0000    0000    0000

Pixel format details:
B: 2 >> 0
G: 3 >> 2
R: 3 >> 5
A: 0 >> 0
Should alpha be honored? No.

Skipping 0 bytes
Resolution: 1448x1072 @ depth 8 (invert: no)
No conversion needed
Now writing PNG file (compression 4)```

**Comment by NiLuJe** [comment]
Huh. Well, I'll be damned. Yet another shitty pixel format. RGB332 is a nasty mad thing that exists.

Since fbgrab runs, I wouldn't mind an actual grab of the *stock* reader displaying stuff with color (ideally, something w/ pure black, pure white, and a few colors. You can even go with a full color wheel if you feel like it ;p).

That should help wrap my head around how it works in practice...

**Comment by NiLuJe** [comment]
So, the good news is that there isn't any black magic involved in the Color handling on the software side.

The bad news is that we don't actually support RGB332 ;p. And it's the kind of thing that's probably going to be highly annoying to implement blind ^^.

**Comment by Teteros** [comment]
@NiLuJe
I've made a test image and the source image, a DSLR shot for context, a screenhot taken with the stock utility of the image viewer and fbgrab of it, all of it attached in the .zip as it would make this message too long.

Sadly it seems fbgrab wasn't able to take a color 24bpp shot, it could only take the grayscale 8 bit PNG but at least the internal sshot tool is able to take color screenshots though that doesn't dump any data of course.
With only `-b 24` it tried to take a 4294966812x1072 shot which failed then with `-b 24 -w 1448` the mmap failed.
Also curiously it takes the fb dumps in landscape 1448x1072 rather than 1072x1448

[kaleido_testimgs.zip](https://github.com/koreader/koreader/files/5103215/kaleido_testimgs.zip)

It's interesting how the stock screenshot utility seems to not have any of the "wavyness" that the rgb filter does to the image in person, fbgrab is more representative there, even if without color.

**Comment by NiLuJe** [comment]
Yeah, I was going to say that the fbgrab dump was going to be weirdly grayscaled, because it assumes grayscale means grayscale, not RGB332 ;p.

(The landscape thing is perfectly normal: remember the `isAlwaysLandscape` trickery: same general idea ;p).

So, in addition, I'll request an actual manual dump of the fb:

`cat /dev/fb0 > fbdump.raw`

That one will actually be handy to see how it looks on the inside ;).

**Comment by NiLuJe** [comment]
Other than that, that's extremely helpful, thanks ;).

And it answers another point I was going to be touching on: you'll notice that the stock software is doing ordered dithering, which is even more necessary than usual here, because of the limited color depth of this pixel format...

**Comment by Teteros** [comment]
FB dump with the test img in focus.
[fbdump.raw.zip](https://github.com/koreader/koreader/files/5103420/fbdump.raw.zip)

**Comment by NiLuJe** [comment]
Thanks!

The good news is that this makes forgetting about color and keeping to grayscale entirely transparent on our end, which is nice, because handling that pixel format properly for color would not be entirely trivial...

**Comment by Teteros** [comment]
> ffmpeg might be able to make heads or tails out of RGB332, if I remember how the hell the raw parser works :D

Out of curiosity I did `ffmpeg -s 1448x1072 -pix_fmt rgb8 -i fbdump.raw fb.png` and it does produce the image (well... not quite, maybe garbage data or wrong pixel format still) but at least it confirms the fb has the testimg
![fb](https://user-images.githubusercontent.com/5067989/90778626-c4ceb480-e2f4-11ea-9a68-dad8b8b5c294.png)

**Comment by NiLuJe** [comment]
Yep, no real RGB332 support in either ffmpeg or IM :/.

**Comment by NiLuJe** [comment]
A quick attempt at doing the conversion [myself](http://trac.ak-team.com/trac/changeset/17722/niluje) didn't yield any better results that that, FWIW.

**Comment by nerd65536** [comment]
The conversion is not a simple pixel format encoding. `-pix_fmt gray8` is probably correct, and if you were to send that grayscale image to the display, it would indeed appear in color.

# Color Filter Array (CFA)
_Every_ addressable pixel of the display has a colored filter on top of it. Each color is painted on the screen in diagonal stripes with slope 1/3.
(Each 3x3 block of pixels has 3 pixels of each color)

Here's a landscape version of how those stripes appear: `CFA.png`
![CFA](https://user-images.githubusercontent.com/7046806/91259799-0bbd0f80-e724-11ea-8612-bb77acd8c69a.png)

If you multiply the `-pix_fmt gray8` version of `fbdump.raw` by the `CFA.png` (in ordinary RGB space), you get an image that simulates how the display appears in real life.
(generated via a "multiply" layer in Gimp)

![CFA_fb](https://user-images.githubusercontent.com/7046806/91261071-89811b00-e724-11ea-96e4-ac1f31099435.png)

We can use knowledge of the CFA image to generate novel images in this format: Multiply the image by the CFA pattern, then convert to grayscale by taking the maximum of each pixel's red, green, and blue values.

For example, the color content of these three images are all rendered essentially identically when shown full-screen (portrait mode, 100% zoom):
[CFA color bars.zip](https://github.com/koreader/koreader/files/5127857/CFA.color.bars.zip)

**Comment by NiLuJe** [comment]
@nerd65536: Huh. Okay, yeah, that makes more sense, thanks ;).

Not *quite* sure how that translates the other way around, though (i.e., RGB32 -> CFA-Gray8).

**Comment by Teteros** [comment]
> if you were to send that grayscale image to the display, it would indeed appear in color.

You're right, can confirm this.
If I take the raw framebuffer dump and convert it to a portrait graycale 8bpp file
`ffmpeg -s 1448x1072 -pix_fmt gray8 -i fbdump.raw -vf transpose=1 gray8.bmp`
Then open it up on the device with `picviewer.app gray8.bmp` it does indeed show up with color, not grayscale when viewed elsewhere.

Of course the grayscale file has to come striped, like the device would render it, any ordinary gray8 pixfmt image will still show up grayscale without it being striped/filtered like the fbdump which leaves the question how to implement the CFA pattern overlay on RGB colored content as the device will probably do the right thing when its passed that format.

**Comment by nerd65536** [comment]
## `fbdump.raw` -> colorized preview
`ffmpeg -s 1448x1072 -pix_fmt gray8 -i fbdump.raw -i CFA.png -filter_complex "format=rgb24, blend=all_mode='multiply'" cfa-rgb.png`

## Arbitrary 1448x1072 RGB image -> CFA-Gray8
I suppose code can explain the conversion more clearly than I can:
`ffmpeg -s 1448x1072 -i input_image.png -i CFA.png -filter_complex "format=rgb24, blend=all_mode='multiply', colorchannelmixer=1:1:1:0:1:1:1:0:1:1:1:0, format=gray8" CFA-Gray8.png`

Optimization:
The CFA itself is a simple pattern where the red, green or blue color corresponds to the image coordinate: `(x/3+y)%3` (within some permutation of off-by-one or sign errors).
So the conversion from rgb24 to CFA-Gray8 might be implemented more directly as: Take the 8-bit r, g, or b value from whichever channel corresponds to that `(x/3+y)%3` mapping.

**Comment by Teteros** [comment]
To add to the above, the direction of the stripes in the pattern is significant.
Since `isAlwaysPortrait = yes` is currently set in this device's definition, the `CFA.png` pattern from https://github.com/koreader/koreader/issues/6479#issuecomment-680678467 would have to be rotated 90 deg clockwise before being multiplied on a portrait image.

This displays colors pretty much identically to the source RGB24 image on the physical device:
```
ffmpeg -i kaleido_1448x1072_24bpp_testimg_src.png -i CFA.png -filter_complex "blend=all_mode=multiply,colorchannelmixer=1:1:1:0:1:1:1:0:1:1:1:0,format=gray8,transpose=1" CFA-Gray8_1072x1448.png
```
![CFA-Gray8_1072x1448](https://user-images.githubusercontent.com/5067989/91345913-03bea780-e7d8-11ea-8839-c12f5ec3a198.png)
when viewing through the build-in PocketBook image viewer.

The color is lost if this grayscale image is rotated in the viewer of course so for making web images for the PB633 one would have to make seperate versions for `1072x1448 (portrait)` and `1448x1072 (landscape)`

**Comment by Frenzie** [comment]
So as not to have to reread all of the above, the device is now supported but only in black & white, isn't it?

**Comment by NiLuJe** [comment]
@Frenzie : Yep!

**Comment by ezdiy** [comment]
I don't have the device, I'm going only by the posted examples courtesy of @nerd65536 and @Teteros

To test this, fetch https://github.com/koreader/koreader-base/blob/45aa1393ee5d26c0dca41d470fcee90cda5f651b/ffi/blitbuffer.lua and https://github.com/koreader/koreader/blob/0ff2a74c8d3175c00c634ad448c92c2d3766f48e/frontend/device/pocketbook/device.lua (or subsequent PR updates) and place into ffi/ and frontend/device/pocketbook

You may need to juggle with https://github.com/koreader/koreader-base/blob/45aa1393ee5d26c0dca41d470fcee90cda5f651b/ffi/blitbuffer.lua#L1017 to get CFA aligned correctly.

**Comment by Teteros** [comment]
@ezdiy Thank you for your rebase yesterday! (Apologies for the delay in testing this as well)

I've applied your koreader (https://github.com/ezdiy/koreader/commit/645fec5a2b8ab5d9657b5126d26b2c32f344582c) and koreader-base (https://github.com/koreader/koreader-base/commit/32ea8a04874baf5147f7fcafad618a73ae17182c)
patches on top of git master and the CFA filter is misaligned indeed, meaning colour don't show up.

Side by side, the source gray8_CFA file that shows in colour on the left and patched koreader on the right from fb dumps:
![CFA_Comparison_Image](https://user-images.githubusercontent.com/5067989/94559001-819f2400-0258-11eb-84d0-6cfc432fc4ff.png)

It looks like the koreader skew is about 45deg while the device expects it tilted more clockwise?
Any suggestions for the values to use in blitbuffer.lua?

**Comment by ezdiy** [comment]
@Teteros Yep, the problem is indeed the pattern itself, not its align. Thanks for testing! The skew was assumed 45 deg, but clearly it isn't - the color column shifts by 1 pixel right *every 3* rows, not *every 1* (when looking at portrait). I'll update the patch to account for that, and then we'll hopefully find the correct settings.

**Comment by ezdiy** [comment]
Should be fixed.

Hopefuly the code - https://github.com/ezdiy/koreader-base/blob/pocketbook_color/ffi/blitbuffer.lua#L1002 - is a bit more human readable now, perhaps to a point a real device owner could bend it to display things properly. The code is horribly inefficient to a point its barely useable on a device. This is for the sake of readability - once we have a working prototype, writing optimized versions is comparably trivial and this "academic" code can be kept as unit test.

It's also now possible to emulate CFA from SDL via

```
EMULATE_BB_TYPE=BB8 ./kodev run
```

**Comment by Teteros** [comment]
The pattern and skew is correct now, good job.

Here's a photo comparison of the device between stock PB picviewer (left) and koreader with new CFA blitter (right)
![photo_pbL_koreaderR](https://user-images.githubusercontent.com/5067989/94748271-d46ffd00-0378-11eb-834f-0dd98de0da3d.jpg)
as well as the framebuffer dumps picviewer.app (bottom), koreader (top)
![fb_pbB_koreaderT](https://user-images.githubusercontent.com/5067989/94748365-0bdea980-0379-11eb-9dce-373da07c89f9.png)

The thing that stands out the most is the banding in the color wheel, although koreader's gray gradient seems smoother.
Performance is significantly worse with color mode on as you noted indeed but still usable.

**Comment by ezdiy** [comment]
Very nice hopefuly some base support will be in few days in nightly builds. There are still limits to this - compared to images, color pdfs will have issues for a while longer due to some of our reflow filtering code being B&W only.

**Comment by rjd22** [comment]
@Frenzie color support for this device has been confirmed working in the next release. This can be closed.

See https://github.com/koreader/koreader/discussions/8818

**Comment by Frenzie** [comment]
Fixed by #8727.
